tsfc/kernel_interface/themis.py

```diff
@@ -0,0 +1,411 @@
+            if isinstance(node, gem.Variable) and node.name.startswith("rt_"):
+                tabulations[node.name] = node.shape
+        self.tabulations = tuple(sorted(tabulations.items()))
+        self.kernel.tabulations = tuple(sorted(tabulations.items()))
+
+    def require_cell_orientations(self):
+        """Set that the kernel requires cell orientations."""
+        self.kernel.oriented = True
+
+    def construct_kernel(self, name, body, quadrature_rule):
+        """Construct a fully built :class:`Kernel`.
+
+        This function contains the logic for building the argument
+        list for assembly kernels.
+
+        :arg name: function name
+        :arg body: function body (:class:`coffee.Block` node)
+        :arg quadrature rule: quadrature rule
+        :returns: :class:`Kernel` object
+        """
+        args = [self.local_tensor, self.coordinates_arg]
+        if self.kernel.oriented:
+            args.append(cell_orientations_coffee_arg)
+        args.extend(self.coefficient_args)
+
+        if self.kernel.integral_type in ["exterior_facet", "exterior_facet_vert"]:
+            args.append(coffee.Decl("unsigned int",
+                                    coffee.Symbol("facet", rank=(1,)),
+                                    qualifiers=["const"]))
+        elif self.kernel.integral_type in ["interior_facet", "interior_facet_vert"]:
+            args.append(coffee.Decl("unsigned int",
+                                    coffee.Symbol("facet", rank=(2,)),
+                                    qualifiers=["const"]))
+
+        for name_, shape in self.tabulations:
+            args.append(coffee.Decl(self.scalar_type, coffee.Symbol(
+                name_, rank=shape), qualifiers=["const"]))
+
+        self.kernel.quadrature_rule = quadrature_rule
+
+        self.kernel.ast = KernelBuilderBase.construct_kernel(self, name, args, body)
+        return self.kernel
+
+    def construct_empty_kernel(self, name):
+        """Return None, since Firedrake needs no empty kernels.
+
+        :arg name: function name
+        :returns: None
+        """
+        return None
+
+
+def prepare_coefficient(coefficient, name, scalar_type, interior_facet=False):
+    """Bridges the kernel interface and the GEM abstraction for
+    Coefficients.
+
+    :arg coefficient: UFL Coefficient
+    :arg name: unique name to refer to the Coefficient in the kernel
+    :arg interior_facet: interior facet integral?
+    :returns: (funarg, expression)
+         funarg     - :class:`coffee.Decl` function argument
+         expression - GEM expression referring to the Coefficient
+                      values
+    """
+    assert isinstance(interior_facet, bool)
+
+    if coefficient.ufl_element().family() == 'Real':
+        # Constant
+
+        # Here we use arrays rather than pointers to fit with how Themis calls kernels
+        funarg = coffee.Decl(scalar_type, coffee.Symbol(name, rank=(1,)), qualifiers=["const"])
```

Comment:
Can you show the difference in the code you want from what you get? Because I'm not sure we will have this control of the argument list with loopy.

This makes:
```c
const double foo[1];
```
Rather than the current 
```c
const * restrict foo
```
What is wrong with the latter for you?

Reply:
I have refactored Themis to use the same kernel arguments as Firedrake (plus tabulations), so the Themis interface is no longer needed.